Mar17 blog (#1100)
Azure/azure-sdk: _posts/2020-03-17-march-release.md

---
title: Azure SDK Releases (March 2020)
layout: post
date: 17 mar 2020
sidebar: releases_sidebar
repository: azure/azure-sdk
---

Welcome to the March release of the Azure SDK.  We have updated the following libraries:

* App Configuration (Java only).
* Event Hubs.
* KeyVault Certificates, Keys, and Secrets v4.0.
* Storage Blobs, File Shares, and Queues.
* Storage Data Lake (new GA release).

These are ready to use in your production applications.  You can find details of all released libraries on [our releases page](https://azure.github.io/azure-sdk/releases/latest/).

New preview releases:

* Azure Identity.
* Cognitive Search.
* Cosmos (Java only).
* Key Vault v4.1
* Text Analytics.

> In addition, we have released a new preview for the Java distributed tracing client, and a new GA release for the Python distributed tracing client.  The distributed tracing clinet allows you to trace a request from the SDK entry-point through to the service using Azure Monitor.

Reviewer: ```suggestion
In addition, we have released a new preview for the Java distributed tracing client, and a new GA release for the Python distributed tracing client.  The distributed tracing client allows you to trace a request from the SDK entry-point through to the service using Azure Monitor.
```

> We believe these are ready for your use, but not yet ready for production.  Between now and the GA release, these libraries may undergo API changes.  We'd love your feedback!  If you use these libraries and like what you see, or you want to see changes, let us know in the GitHub issues for the appropriate language. 

Reviewer: ```suggestion
We believe these are ready for you to use and experiment with, but not yet ready for production.  Between now and the GA release, these libraries may undergo API changes.  We'd love your feedback!  If you use these libraries and like what you see, or you want to see changes, let us know in GitHub issues. 
```

"ready for your use" felt kinda awkward to me. The last part seems obvious.

## Getting Started

Use the links below to get started with your language of choice.  You will notice that all the preview libraries are tagged with "preview".

* [.NET release notes]({{site.baseurl}}/releases/2020-02/dotnet.html)
* [Java release notes]({{site.baseurl}}/releases/2020-02/java.html)
* [Python release notes]({{site.baseurl}}/releases/2020-02/python.html)
* [JavaScript release notes]({{site.baseurl}}/releases/2020-02/js.html)

If you want to dive deep into the content, the release notes linked above and the change logs they point to give more details on what has changed.

## Introducing Azure Cognitive Search

> This month, we are introducing a preview of the Azure Cognitive Search client.  [Azure Cognitive Search](https://azure.microsoft.com/services/search) is search-as-a-service, allowing developers to add a rich search experience over private, heterogenous content in web, mobile, and enterprise applications.  You've probably seen this type of search experience in action when you use a product search capability within an eCommerce site.  Let's take a look at how you can implement the search capability in your own client applications.  For this demonstration, I'm going to be using JavaScript and the React framework.

Reviewer: ```suggestion
This month, we are introducing a preview of the Azure Cognitive Search client.  [Azure Cognitive Search](https://azure.microsoft.com/services/search) is search-as-a-service, allowing developers to add a rich search experience over private, heterogenous content in web, mobile, and enterprise applications.  You've probably seen this type of search experience in action when you use a product search capability within an e-commerce site.  Let's take a look at how you can implement the search capability in your own client applications.  For this demonstration, I'm going to be using JavaScript and the React framework.
```

I've not seen eCommerce before. Reminds me of eMachines 😂

Start by creating a singleton service client:

```javascript
import { SearchIndexClient, SearchApiKeyCredential } from '@azure/search';
import searchClientConfiguration from './searchConfig.json';

const searchClient = new SearchIndexClient(
  searchClientConfiguration.endpoint,
  searchClientConfiguration.indexName,
  new SearchApiKeyCredential(searchClientConfiguration.apiKey)
);

export default searchClient;
```

Now, let's imagine a search page with a search box at the top and a list of results:

```javascript
import React, { useState } from 'react';
import searchClient from './searchClient';

> const SearchPage = () => {

Author: Yes, I did.  Agreed on the CORS bit.  I think it's reasonable for a client app.  I'll add a note here to indicate what the guidance is.

> const SearchPage = () => {

Reviewer 2: Very cool that this scenario "just works" with the new SDK!

[Facets](https://docs.microsoft.com/azure/search/search-filters-facets) allow you to display buckets that assist the user to further refine their search.  If you are search for a chair, you might want to refine your search based on location, rating, or color.  Here is an example component:

```javascript
import React, { useEffect, useState } from 'react';
import searchClient from './searchClient';

const SearchPage = ({ searchString }) => {
  const [locations, setLocations] = useState([]);
  const [rating, setRating] = useState([]);

  useEffect(() => {
    const runFacetQuery = async () => {
      const response = await searchClient.search({
        searchText: searchString,
        facets: [

>           "Location,count:3,sort:count",

Reviewer 2: nit: It's a bit unusual to have mixed-case fields. Typically they'd either all be `camelCase` or `PacalCase` (note that field names are case-sensitive).

* Event Hubs.
* KeyVault Certificates, Keys, and Secrets v4.0.
* Storage Blobs, File Shares, and Queues.
* Storage Data Lake (new GA release).

These are ready to use in your production applications.  You can find details of all released libraries on [our releases page](https://azure.github.io/azure-sdk/releases/latest/).

New preview releases:

* Azure Identity.
* Azure Cognitive Search.
* Cosmos (Java only).
* Key Vault v4.1
* Text Analytics.

In addition, we have released a new preview for the Java distributed tracing client, and a new GA release for the Python distributed tracing client.  The distributed tracing client allows you to trace a request from the SDK entry-point through to the service using Azure Monitor.

We believe these are ready for you to use and experiment with, but not yet ready for production.  Between now and the GA release, these libraries may undergo API changes.  We'd love your feedback!  If you use these libraries and like what you see, or you want to see changes, let us know in GitHub issues. 

## Getting Started

Use the links below to get started with your language of choice.  You will notice that all the preview libraries are tagged with "preview".

> * [.NET release notes]({{site.baseurl}}/releases/2020-02/dotnet.html)

Reviewer 3: These all point at Feb relnotes files. I'm assuming they should all be 2020-03

Reviewer 4: (Btw, they all return 404s)

Author: Will update the march.  They return 404s because they are relative links - not absolute.

```javascript
import React, { useEffect, useState } from 'react';
import searchClient from './searchClient';

const SearchPage = ({ searchString }) => {
  const [locations, setLocations] = useState([]);
  const [rating, setRating] = useState([]);

  useEffect(() => {
    const runFacetQuery = async () => {
      const response = await searchClient.search({
        searchText: searchString,
        facets: [
          "location,count:3,sort:count",
          "rating,count:5,sort:count",
          "color,count:3,sort:count"
        ]
      });

      setLocations(facets.location.map(v => v.value));
      setRating(facets.rating.map(v => v.value));
    };

    runFacetQuery();
  });

  return (
    <>
      <SearchInputBox onSearch={(value) => searchProducts(value)} />
      <FacetDisplay search={searchString} />
      <div className="searchResult">
        <h2>Location</h2>
        <ul>
          {locations.map(result => <li key={result}>{result}</li>)}
        </ul>
        <h2>Rating</h2>
        <ul>
          {rating.map(result => <li key={result}>{result}</li>)}
        </ul>
      </div>
    </>
  );
};
```

> Although both the search page and the facet display components both do searches, you can combine the searches into one operation.  Once the user clicks on a facet, you will want to refine the search using an OData search filter.  For example, let's say the user clicked on the `Location=US` facet.  You can perform that search as follows:

Reviewer 5: In this phrase "Although both the search page and the facet display components both do searches" -- there are two instances of both, can you get rid of one?

Author: I show off the combined search next.  The point of it was to show them separated, and then bring them together.

```javascript
import { odata } from '@azure/search';

// In this case $facetLocation = 'US'
const response = await searchClient.search({
  searhcText: searchString,
  filter: odata`location eq ${facetLocation}`,
  orderBy: [ "price desc" ],
  select: [ "productName", "price" ],
  top: 20,
  skip: 0
  facets: [
    "location,count:3,sort:count",
    "rating,count:5,sort:count",
    "color,count:3,sort:count"
  ]
});
```

The `odata` formatter ensures that the variables you use are quoted properly.

Let's turn our attention to the search input box.  One of the features of any good product search is an autocompleter.  Azure Cognitive Search provides [suggesters](https://docs.microsoft.com/azure/search/index-add-suggesters) to allow the search service to suggest records you might be interested in.  You can use this to implement an autocomplete feature:

```javascript
const response = await searchClient.autocomplete({
  searchText: searchString,
  suggesterName: 'sg'
});
const suggestions = response.results || [];
```

You can then use the suggestions to populate the drop-down autocomplete box.

> For more information on the Azure Cognitive Search SDK for JavaScript, check out the [API documentation](https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-search/11.0.0-preview.1/index.html).  The Azure Cognitive Search SDK is also available for Python, Java, and .NET.

Reviewer 5: Are you sure about the SDK for Python and Java?  The Management API is available in those languages, but I don't think the Search API is onboard yet.

Reviewer 2: Yes, all four languages are included in this preview. 😊

> For more information on the Azure Cognitive Search SDK for JavaScript, check out the [API documentation](https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-search/11.0.0-preview.1/index.html).  The Azure Cognitive Search SDK is also available for Python, Java, and .NET.

Reviewer 4: package version is not showing up in API link

Author: Not sure what you mean.  The link works.

> For more information on the Azure Cognitive Search SDK for JavaScript, check out the [API documentation](https://azuresdkdocs.blob.core.windows.net/$web/javascript/azure-search/11.0.0-preview.1/index.html).  The Azure Cognitive Search SDK is also available for Python, Java, and .NET.

Reviewer 4: Maybe we could give the links to these other SDKs?

Author: They aren't live yet, so I can't just yet.

This month, we are introducing a preview of the Azure Cognitive Search client.  [Azure Cognitive Search](https://azure.microsoft.com/services/search) is search-as-a-service, allowing developers to add a rich search experience over private, heterogenous content in web, mobile, and enterprise applications.  You've probably seen this type of search experience in action when you use a product search capability within an e-commerce site.  Let's take a look at how you can implement the search capability in your own client applications.  For this demonstration, I'm going to be using JavaScript and the React framework.

> > We recommend that most application use an intermediary web API service to protect the API key.  You can write your web API using Azure Functions and Node.js.  The same JavaScript API is used to access Azure Cognitive Search.

Reviewer 2: ```suggestion
> We recommend that most applications use an intermediary web API service to protect the API key.  You can write your web API using Azure Functions and Node.js.  The same JavaScript API is used to access Azure Cognitive Search.
```

// In this case $facetLocation = 'US'
const response = await searchClient.search({
  searhcText: searchString,
  filter: odata`location eq ${facetLocation}`,

>   orderBy: [ "price desc" ],

Reviewer 2: ```suggestion
  orderBy: [ "Price desc" ],
```

Author: I just changed everything to be camel-case :-)

Reviewer 2: That works too :)

>   select: [ "productName", "price" ],

Reviewer 2: ```suggestion
  select: [ "ProductName", "Price" ],
```
